network/p2p/dns/resolver_test.go

```diff
@@ -83,10 +87,51 @@ func TestResolver_Error(t *testing.T) {
 	// this evaluates that upon returning an error, the result is not cached, so the next invocation again goes
 	// through the resolver.
 	resolverWG := mockBasicResolverForDomains(t, &basicResolver, ipTestCase, txtTestCases, !happyPath, times)
-	queryWG := queryResolver(t, times, resolver, txtTestCases, ipTestCase, !happyPath)
+	queryWG := syncThenAsyncQuery(t, times, resolver, txtTestCases, ipTestCase, !happyPath)
 
 	unittest.RequireReturnsBefore(t, resolverWG.Wait, 1*time.Second, "could not resolve all expected domains")
 	unittest.RequireReturnsBefore(t, queryWG.Wait, 1*time.Second, "could not perform all queries on time")
+
+	// since resolving hits an error, cache is invalidated.
+	require.Empty(t, resolver.c.ipCache)
+	require.Empty(t, resolver.c.txtCache)
+
+	unittest.RequireCloseBefore(t, resolver.Done(), 10*time.Millisecond, "could not stop dns resolver on time")
+}
+
+// TestResolver_Expired_Invalidated evaluates that when resolver is queried for an expired entry, it returns the expired one, but queries async on the
+// network to refresh the cache. However, when the query hits an error, it invalidates the cache.
+func TestResolver_Expired_Invalidated(t *testing.T) {
+	basicResolver := mocknetwork.BasicResolver{}
+	resolver := NewResolver(metrics.NewNoopCollector(),
+		WithBasicResolver(&basicResolver),
+		WithTTL(1*time.Second)) // 1 second TTL for test
+
+	unittest.RequireCloseBefore(t, resolver.Ready(), 10*time.Millisecond, "could not start dns resolver on time")
+
+	// one test case for txt and one for ip
+	txtTestCases := txtLookupFixture(1)
+	ipTestCase := ipLookupFixture(1)
+
+	// mocks test cases cached in resolver and waits for their expiry
+	mockCacheForDomains(resolver, ipTestCase, txtTestCases)
+	time.Sleep(1 * time.Second)
+
+	// queries for an expired entry must return the expired entry but also fire an async update on it.
+	// though we mock async update to fail, so the cache should be invalidated literally.
+	// mocks underlying basic resolver invoked once per domain and returns an error on each domain
+	resolverWG := mockBasicResolverForDomains(t, &basicResolver, ipTestCase, txtTestCases, !happyPath, 1)
+	// queries are answered by cache, so resolver returning an error only invalidates the cache asynchronously for the first time.
+	queryWG := syncThenAsyncQuery(t, 1, resolver, txtTestCases, ipTestCase, happyPath)
+
+	unittest.RequireReturnsBefore(t, queryWG.Wait, 1*time.Second, "could not perform all queries on time")
+	unittest.RequireReturnsBefore(t, resolverWG.Wait, 1*time.Second, "could not resolve all expected domains")
+
+	// since resolving hits an error, cache is invalidated.
+	require.Empty(t, resolver.c.ipCache)
+	require.Empty(t, resolver.c.txtCache)
```

Comment:
`invalidateIPCacheEntry` (resp `invalidateTXTCacheEntry`) is called in a `unit.Launch`, which means you haven't captured an happens-after by waiting for `queryWG` & `resolverWG`, and that these 2 `require.Empty` are in a data race with the invalidations.

Reply:
This one is still present, and I suspect `TestResolver_CacheExpiry` exhibits the issue as well. A quick fix would be to put the `require.Empty` in an eventually.

Reply:
Moved `require.Empty` statements after the invocation of `resolver.Done`. The `Done` method returns once all outstanding concurrent `unit.Launch` operations terminate. So by the time `resolver.Done` returns we are sure that the termination of `invalidateIPCacheEntry`, `invalidateTXTCacheEntry`, as well as any other method invoked in `unit.Launch`  happens-before the `unit.Done` termination. So, evaluating `require.Empty` right after `Done` makes it concurrency safe.